bpf/process/types/basic.h

```diff
@@ -848,7 +845,7 @@ FUNC_INLINE void copy_reverse(__u8 *dest, uint len, __u8 *src, uint offset)
 	// Alternative (prettier) fixes resulted in a confused verifier
 	// unfortunately.
 	for (i = 0; i < (STRING_POSTFIX_MAX_MATCH_LENGTH - 1); i++) {
-		dest[i & STRING_POSTFIX_MAX_MASK] = src[(len + offset - 1 - i) & PATH_MASK];
+		dest[i & STRING_POSTFIX_MAX_MASK] = src[(len + offset - 1 - i) & (BINARY_PATH_MAX_LEN - 1)];
```

Comment:
So this is more restrictive so it should be fine, and postfix on strings is also limited IIRC, would be nice to have @kevsecurity ack on that. BINARY_PATH_MAX_LEN is 256 which is < 128 (see below).

```C
#define STRING_POSTFIX_MAX_LENGTH 128
#define STRING_POSTFIX_MAX_MASK	  (STRING_POSTFIX_MAX_LENGTH - 1)
#ifdef __LARGE_BPF_PROG
#define STRING_POSTFIX_MAX_MATCH_LENGTH STRING_POSTFIX_MAX_LENGTH
#else
#define STRING_POSTFIX_MAX_MATCH_LENGTH 95
#endif
```

Reply:
I'd rather leave this at PATH_MASK if possible as it better represents the length of a path on Linux. Currently we appear to limit paths (when copying) to 256 bytes so this change would almost certainly work. But if we later increase path lengths to 4096, then this will also need changing.

Reply:
Okay, I can see that this doesn't work for the execve_map_value->bin.path because it is only BINARY_PATH_MAX_LENGTH in size. But string/charbuf buffers could be 4096 in size, and postfix matching needs to work on these too. I suggest a file_copy_reverse function that is used only in file/binary copying/matching, and maybe make binary postfix matching a LARGE_BPF_PROG option only to avoid hitting complexity limits on older kernels.

Reply:
The reason I changed `PATH_MASK` is because verifier check failed. AFAIK,  `copy_reverse` is used only in one place (except new added call):

https://github.com/cilium/tetragon/blob/95f263729f129bb1b2d02ae157b9b7d24804832c/bpf/process/types/basic.h#L878

The verification in this case is OK, because it is enough place to hold the source buffer:

https://github.com/cilium/tetragon/blob/95f263729f129bb1b2d02ae157b9b7d24804832c/bpf/lib/generic.h#L50

I understand correctly, I suggest to create a function  `file_copy_reverse`, that like `copy_reverse`, but with (BINARY_PATH_MAX_LEN - 1) mask? If so, maybe rename `copy_reverse` to `arg_copy_reverse` than?

Reply:
I found the answer [here](https://github.com/cilium/tetragon/pull/2689#pullrequestreview-2191352985), thanks:).

Reply:
> AFAIK, `copy_reverse` is used only in one place (except new added call):
> 
> https://github.com/cilium/tetragon/blob/95f263729f129bb1b2d02ae157b9b7d24804832c/bpf/process/types/basic.h#L878

Except that handles character buffers up to 4096 chars in length.

> The verification in this case is OK, because it is enough place to hold the source buffer:
> 
> https://github.com/cilium/tetragon/blob/95f263729f129bb1b2d02ae157b9b7d24804832c/bpf/lib/generic.h#L50
> 
> If I understand correctly, you suggest to create a function `file_copy_reverse`, that like `copy_reverse`, but with (BINARY_PATH_MAX_LEN - 1) mask? If so, maybe rename `copy_reverse` to `arg_copy_reverse` than?

Yep. Then char buffers can be much longer and still do postfix matching correctly, without affecting the masking needed for binary postfix matching.